stdlib/logging/__init__.pyi

```diff
@@ -63,10 +63,16 @@ if sys.version_info >= (3, 11):
 if sys.version_info >= (3, 12):
     __all__ += ["getHandlerByName", "getHandlerNames"]
 
+class _SupportsFilter(Protocol):
+    if sys.version_info >= (3, 12):
+        def filter(self, record: LogRecord) -> bool | LogRecord: ...
+    else:
+        def filter(self, record: LogRecord) -> bool: ...
+
 _SysExcInfoType: TypeAlias = tuple[type[BaseException], BaseException, TracebackType | None] | tuple[None, None, None]
 _ExcInfoType: TypeAlias = None | bool | _SysExcInfoType | BaseException
 _ArgsType: TypeAlias = tuple[object, ...] | Mapping[str, object]
-_FilterType: TypeAlias = Filter | Callable[[LogRecord], bool]
+_FilterType: TypeAlias = _SupportsFilter | Callable[[LogRecord], bool]
```

Comment:
- Turns out we should keep `Filter` here, based on the mypy_primer output. Sphinx uses `# type: ignore` when they define a custom filter with a weird `filter()` method. Including `Filter` here means that they don't get errors when adding the custom filter, only when defining it. Getting errors in only one place is better IMO.
- On 3.12+, we should probably support the `LogRecord` return type here too.

```suggestion
if sys.version_info >= (3, 12):
    _FilterType: TypeAlias = Filter | _SupportsFilter | Callable[[LogRecord], bool | LogRecord]
else:
    _FilterType: TypeAlias = Filter | _SupportsFilter | Callable[[LogRecord], bool]
```

Reply:
Commit https://github.com/python/typeshed/pull/11018/commits/eed6015703ce4369bc4d032c33fa768e53ce58f0 🤓